c_src/device/cairo.c

```diff
@@ -0,0 +1,399 @@
+#include <cairo.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/fb.h>
+#include <stdlib.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include "cairo_ctx.h"
+#include "comms.h"
+#include "device.h"
+#include "fontstash.h"
+#include "ops/script_ops.h"
+
+const char* device = "/dev/fb0";
+
+typedef struct {
+  int fd;
+
+  struct fb_var_screeninfo var;
+  struct fb_fix_screeninfo fix;
+} cairo_fb_t;
+
+cairo_fb_t g_cairo_fb = {0};
+
+extern device_info_t g_device_info;
+extern device_opts_t g_opts;
+
+// Support for 8 bits per pixel colors
+unsigned short red[256], green[256], blue[256];
+struct fb_cmap map332 = {0, 256, red, green, blue, NULL};
+unsigned short red_b[256], green_b[256], blue_b[256];
+struct fb_cmap map_back = {0, 256, red_b, green_b, blue_b, NULL};
+
+void make332map(struct fb_cmap *map)
+{
+  int rs, gs, bs, i;
+  int r = 8, g = 8, b = 4;
+
+  map->red = red;
+  map->green = green;
+  map->blue = blue;
+
+  rs = 256 / (r - 1);
+  gs = 256 / (g - 1);
+  bs = 256 / (b - 1);
+
+  for (i = 0; i < 256; i++) {
+    map->red[i]   = (rs * ((i / (g * b)) % r)) * 255;
+    map->green[i] = (gs * ((i / b) % g)) * 255;
+    map->blue[i]  = (bs * ((i) % b)) * 255;
+  }
+}
+
+void set8map(int fh, struct fb_cmap *map)
+{
+    if (ioctl(fh, FBIOPUTCMAP, map) < 0) {
+        log_error("cairo: Error putting colormap");
+    }
+}
+
+void get8map(int fh, struct fb_cmap *map)
+{
+    if (ioctl(fh, FBIOGETCMAP, map) < 0) {
+        log_error("cairo: Error getting colormap");
+    }
+}
+
+void set332map(int fh)
+{
+    make332map(&map332);
+    set8map(fh, &map332);
+}
+
+int device_init(const device_opts_t* p_opts,
+                device_info_t* p_info,
+                driver_data_t* p_data)
+{
+  if (g_opts.debug_mode) {
+    log_info("cairo %s", __func__);
+  }
+
+  if ((g_cairo_fb.fd = open(device, O_RDWR)) == -1) {
+    log_error("Failed to open device %s: %s", device, strerror(errno));
+    return -1;
+  }
+
+  if (ioctl(g_cairo_fb.fd, FBIOGET_VSCREENINFO, &g_cairo_fb.var)) {
+    log_error("Failed to get fb_var_screeninfo: %s", strerror(errno));
+    return -1;
+  }
+
+  if (ioctl(g_cairo_fb.fd, FBIOGET_FSCREENINFO, &g_cairo_fb.fix)) {
+    log_error("Failed to get fb_fix_screeninfo: %s", strerror(errno));
+    return -1;
+  }
+
+  scenic_cairo_ctx_t* p_ctx = calloc(1, sizeof(scenic_cairo_ctx_t));
+
+  FT_Error status = FT_Init_FreeType(&p_ctx->ft_library);
+  if (status != 0) {
+    log_error("cairo: FT_Init_FreeType: Error: %d", status);
+    close(g_cairo_fb.fd);
+    free(p_ctx);
+
+    return -1;
+  }
+
+  p_ctx->ratio = 1.0f;
+  p_ctx->dist_tolerance = 0.1f * p_ctx->ratio;
+
+  size_t pix_count = g_cairo_fb.var.xres * g_cairo_fb.var.yres;
+  switch (g_cairo_fb.var.bits_per_pixel)
+  {
+  case 8:
+    p_ctx->fbbuff.c = (uint8_t*)malloc(pix_count * sizeof(uint8_t));
+    break;
+  case 15:
```

Comment:
Which format is this one, OOC?

Reply:
The 15 bit formats are the one that I haven't actually been able to test. I was able to test all the other color depths using `fbset`. The 15 bit formats will be RGB555 or BGR555.

Reply:
Ah cool, thank you.